### DISCUSSION TITLE
Publish AOT with error about Microsoft.WebTools.Shared.Exceptions.WebToolsException

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#14853] [Q&A] [answered]
Recently I wrote an Avalonia program and planned to release it as Linux arm to run on the Raspberry Pi.

If the program is published normally, everything runs fine, but if I publish it using AOT, it prompts:
Publish has encountered an error.
Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details.

A diagnostic log has been written to the following location:
"C:\Users\Admin\AppData\Local\Temp\tmp38FB.tmp"

Then I opened the file tmp38FB.tmp to view the log:
2024/3/7 11:20:14
System.AggregateException: One or more errors occurred. ---> Microsoft.WebTools.Shared.Exceptions.WebToolsException: Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details.
    --- End of inner exception stack trace ---
---> (Inner Exception #0) Microsoft.WebTools.Shared.Exceptions.WebToolsException: Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details. <---

Microsoft.WebTools.Shared.Exceptions.WebToolsException: Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details.

===================

What's the reason why I can't publish using AOT? My experience with AOT publishing is wrong and I don't know how to troubleshoot this issue.

**Comment by maxkatz6** [comment]
What is `Microsoft.WebTools.Shared.Exceptions.WebToolsException`? Do you reference any web related libraries? 
It's not related to Avalonia, so I can't help much. But it looks like this web library doesn't support NativeAOT well.

**Reply by maxkatz6** [reply]
Could it be that you try to publish your app as a web app?
Try with simple `dotnet publish` CLI command first.

**Reply by kyuranger** [reply]
I tried dotnet publish and it published succussed with a windows application.

After I tried to run the application, windows throw this error as below:
The system detected a stack-based buffer overflow in this application. An overflow could allow a malicious user to gain control of this application. This overrun could potentially allow a malicious user to gain control of this application.

**Reply by maxkatz6** [reply, marked as answer]
This is pretty unusual error. If you can make a minimal sample, please consider reporting it to https://github.com/dotnet/runtime.

**Reply by kyuranger** [reply]
What's more, when I try 
dotnet publish -r linux-arm64 -c Release
in https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net7%2Cwindows 
The PowerShell throws this error:
C:\Users\Admin\.nuget\packages\microsoft.dotnet.ilcompiler\8.0.2\build\Microsoft.NETCore.Native.Publish.targets(60,5): error : Cross-OS native compilation is not supported.

**Reply by kyuranger** [reply]
Well, it seems that I can only use non-AOT publishing first. Thank you very much.

**Reply by maxkatz6** [reply]
> Cross-OS native compilation is not supported.

For linux you can only compile on linux with AOT. The same with other OS.
